File: DS_Problems/Array/Patterns/SlidingWindow.cpp
#include <bits/stdc++.h>
using namespace std;

// The Sliding Window algorithm is a technique used to efficiently 
// process arrays or lists in a linear manner, typically to find 
// or compute something within a fixed-size window that "slides" 
// through the array. It is useful for solving problems involving 
// substring, subarray, or subsequence manipulation.

class Solution
{
private:
    vector<int> v;

public:
    Solution(vector<int> _v)
    {
        v = _v;
    }

    void solve(int k)
    {
        int windowSize = k;
        int windowSum = 0, maxSum = 0;

        for(int i=0;i<windowSize;i++){
            windowSum+=v[i];
        }

        for(int i=windowSize;i<v.size();i++){
            windowSum+=v[i]-v[i-windowSize];
            maxSum = max(maxSum,  windowSum);
        }

        cout<<"Max sum of subarray with length k is: "<<maxSum<<endl;
    }
};

int main()
{
    Solution s({1, 3, 2, 4, 6, 2, 1});
    s.solve(3);
}

File: Algo_Problems/SlidingWindow/SlidingWindow.java
package Algo_Problems.SlidingWindow;

/*
 * Sliding Window Technique is a method used to efficiently solve 
 * problems that involve defining a window or range in the input 
 * data (arrays or strings) and then moving that window across the 
 * data to perform some operation within the window. This technique 
 * is commonly used in algorithms like finding subarrays with a 
 * specific sum, finding the longest substring with unique characters, 
 * or solving problems that require a fixed-size window to process 
 * elements efficiently.
 * 
 * Problem:
 * Given an array of integers of size ‘n’, Our aim is to calculate the 
 * maximum sum of ‘k’ consecutive elements in the array.
 */

public class SlidingWindow {
    public static void main(String[] args) {
        int nums[] = { 1, 4, 7, 9, 2, 3, 16 };
        solve(nums, 3);
    }

    static void solve(int nums[], int k) {
        if (nums.length < k) {
            System.out.println("Invalid");
            return;
        }
        int maxSum = 0;

        // First calculate the sum of first window
        for (int i = 0; i < k; i++) {
            maxSum += nums[i];
        }

        // Now shift the window
        int curSum = maxSum;
        for (int i = k; i < nums.length; i++) {
            curSum = nums[i] - nums[i - k] + curSum;
            maxSum = Math.max(maxSum, curSum);
        }

        System.out.println(maxSum);
    }
}
